doc/build/core/temporal.rst

```diff
@@ -0,0 +1,238 @@
+way as usual: either with the ``primary_key`` argument to :class:`Column`, or
+with a :class:`PrimaryKeyConstraint` or :class:`UniqueConstraint` table
+argument. The below is an example:
+
+
+.. code-block:: python
+
+    from sqlalchemy import Period, UniqueConstraint
+
+    t = Table(
+        "t",
+        metadata,
+        Column("id", Integer, primary_key=True),
+        Column("name", String(30)),
+        Column("tstart", DATE),
+        Column("tend", DATE),
+        Period("my_period", "tstart", "tend", primary_key=True),
+        UniqueConstraint("name", "my_period")
+    )
+
+Generated DDL:
+
+.. code-block:: sql
+
+    CREATE TABLE t (
+        id INTEGER,
+        name VARCHAR(30),
+        tstart DATE,
+        tend DATE,
+        PERIOD FOR my_period (tstart, tend),
+        PRIMARY KEY (id, my_period WITHOUT OVERLAPS),
+        UNIQUE (name, my_period WITHOUT OVERLAPS)
+    )
+
+Nothing special is needed to specify `WITHOUT OVERLAPS`, it is added
+automatically if there is a period as a primary key.
+
+Working with System-Time Periods
+--------------------------------
+
+System time periods are represented in DDL by adding ``WITH SYSTEM VERSIONING``
+to a table definition and by creating a ``PERIOD FOR SYSTEM_TIME(...)``
+construct. Some backends may allow omitting one of these in some cases.
+
+This can be accomplished in SQLAlchemy using the :class:`SystemTimePeriod`
+schema item, as below.
+
+.. code-block:: python
+
+    from sqlalchemy import SystemTimePeriod
+
+    t = Table(
+        "t",
+        metadata,
+        Column("id", Integer),
+        Column("name", String(30)),
+        Column("bigblob", BLOB, system_versioning=False),
+        Column("sys_start", TIMESTAMP),
+        Column("sys_end", TIMESTAMP),
+        SystemTimePeriod("sys_start", "sys_end"),
+    )
+
+SQLAlchemy performs the following steps to create the DDL:
+
+#. Look for a ``SystemTimePeriod`` object. If present, add ``WITH SYSTEM
+   VERSIONING`` to the table's ``CREATE`` statement.
+#. Locate the columns specified for the ``SystemTimePeriod()`` ``start`` and
+   ``end`` arguments (first two positional arguments). For each of these,
+   perform the following:
+
+    * Check if the column has a :class:`Computed` object. If so, no further
+      action is taken.
+    * If not, add a :class:`Computed` object that will render as ``GENERATED
+      ALWAYS AS ROW START`` or ``GENERATED ALWAYS AS ROW END``.
+  If no column arguments are provided to ``SystemTimePeriod()``, ``WITH SYSTEM
+  VERSIONING`` will be added to the table but no ``GENERATED`` columns will be
+  specified. This is useful if the backend supports implicit system versioning
+  columns (e.g. MariaDB).
+
+#. Look for any columns with a ``system_versioning`` argument. This will add
+   ``WITH SYSTEM VERSIONING`` or ``WITHOUT SYSTEM VERSIONING`` to that column.
+   Not all backends support individual column opting in/out.
+
+The resulting DDL is below:
+
+.. code-block:: sql
+    
```

Comment:
**Federico Caselli** (CaselIT) wrote:

leftover space

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3803